echopype/convert/set_groups_ad2cp.py

```diff
@@ -38,126 +30,193 @@ class SetGroupsAd2cp(SetGroupsBase):
+                field = data_record_format.get_field(field_name)
+                if field is None:
+                    field_dimensions = Field.default_dimensions()
+                else:
+                    field_dimensions = field.dimensions(packet.data_record_type)
+
+                    if field_name not in dims:
+                        dims[field_name] = field_dimensions
+                    if field_name not in dtypes:
+                        field_entry_size_bytes = field.field_entry_size_bytes
+                        if callable(field_entry_size_bytes):
+                            field_entry_size_bytes = field_entry_size_bytes(packet)
+                        dtypes[field_name] = field.field_entry_data_type.dtype(
+                            field_entry_size_bytes
                         )
-                coords = {
-                    "ping_time": [packet.timestamp],
-                    ping_time_dim: [packet.timestamp],
-                }
-                if "beams" in packet.data_exclude:
-                    coords["beam"] = packet.data_exclude["beams"]
-                new_packet = xr.Dataset(data_vars=data_vars, coords=coords)
-
-                # modify in place to reduce memory consumption
-                packets[i] = new_packet
-            if len(packets) > 0:
-                packets = merge_attrs(packets)
-                return xr.combine_by_coords(
-                    packets,
-                    data_vars="minimal",
-                    coords="minimal",
-                    combine_attrs="override",
+                    if field_name not in units:
+                        units[field_name] = field.units()
+
+                if field_name in packet.data:  # field is in this packet
+                    fields[field_name].append(packet.data[field_name])
+                    field_exists[field_name] = True
+                else:  # field is not in this packet
+                    # pad the list of field values with an empty array so that
+                    #   the time dimension still lines up with the field values
+                    fields[field_name].append(np.array(0))
+                    pad_idx[field_name].append(len(fields[field_name]) - 1)
+
+        # add dimensions to dims if they were not found
+        for field_name in fields.keys():
+            if field_name not in dims:
+                dims[field_name] = Field.default_dimensions()
+
+        # add dtypes to dtypes if they were not found
+        for field_name in fields.keys():
+            if field_name not in dtypes:
+                dtypes[field_name] = DataType.default_dtype()
+
+        # replace padding with correct shaped padding
+        for field_name, pad_idxs in pad_idx.items():
+            for i in pad_idxs:
+                fields[field_name][i] = np.zeros(
+                    np.ones(len(dims[field_name]) - 1, dtype="u1"),  # type: ignore
+                    dtype=dtypes[field_name],
                 )
```

Comment:
I don't think that you can have this block in 129-137 without 108-118 in the `packet` loop because to get the field dimensions we need the data record format of the `packet`, so we'd have to loop over the packets.

These blocks in 129-137 also can't be deleted because they cover cases where the EchoData group has a variable that doesn't exist in the data records. That might happen if the raw file doesn't have any packets with data records which contain those variables or if the field is created by echopype itself (e.g. from a bitfield).

Reply:
I did combine the loops though

Reply:
I see, ok, thanks for the explanation! This is something we can probably remove during the reorganization of variables into `Beam_groupX` -- like if some variables do not exist, the entire associated `Beam_groupX` may not have to exist. Let's revisit this there.